### PR TITLE
chore: enable `-Wunused`

### PR DESCRIPTION
**Pull request by kpodsiad**
Turned out we had a lot of unused warnings.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -276,6 +271,35 @@ lazy val V = new {
...
+    // desugaring of for yield caused pattern var to complain
+    // https://github.com/scala/bug/issues/10287
+    "-Wconf:msg=parameter value .+ in anonymous function:silent",
+    // silence unused parameters in mtags
+    "-Wconf:src=*.ScaladocParser.scala&msg=in method reportError:silent",
+    "-Wconf:src=*.Completions.scala&msg=in method (isCandidate|isPrioritized):silent",
+    "-Wconf:src=*.JavaMtags.scala&msg=in method (visitConstructor|visitMethod):silent",
+    "-Wconf:src=*.MtagsIndexer.scala&msg=in method visitOccurrence:silent",
+    // silence "The outer reference in this type test cannot be checked at run time." in mtags
+    "-Wconf:src=.*(CompletionProvider|ArgCompletions|Completions|Keywords).scala&msg=The outer reference:silent"
```
Comment (kpodsiad) [inline build.sbt:289]: I need to dive into this warning deeper, but for now, I silenced it to have 0 warnings in the problems tab.

File: build.sbt [hunk 2]
```diff
@@ -276,6 +271,35 @@ lazy val V = new {
     ).toList
 }
 
+// -Xlint is unusable because of
+// https://github.com/scala/bug/issues/10448
+def lintingOptions(scalaVersion: String) = {
+  val unused213 = "-Wunused"
+  val unused3 = "-Wunused:all"
```
Comment (kpodsiad) [inline build.sbt:278]: I think it's better to enable all of them and them eventually filter out noisy ones via `Wconf`

File: build.sbt [hunk 3]
```diff
@@ -276,6 +271,35 @@ lazy val V = new {
...
+  val common = List(
+    // desugaring of for yield caused pattern var to complain
+    // https://github.com/scala/bug/issues/10287
+    "-Wconf:msg=parameter value .+ in anonymous function:silent",
+    // silence unused parameters in mtags
```
Comment (ckipp01) [inline build.sbt:283]: Maybe a silly question, but are they false positives? Or what not remove the params if they are unused instead of silencing them?

File: build.sbt [hunk 4]
```diff
@@ -276,6 +271,35 @@ lazy val V = new {
...
+    "-Wconf:src=*.ScaladocParser.scala&msg=in method reportError:silent",
```
Comment (ckipp01) [inline build.sbt:284]: This seems pretty broad, or is it just me? Like are there more messages that the target one that include `in method` that we might be silencing that we don't want to?

File: metals/src/main/scala/scala/meta/internal/metals/codelenses/CodeLens.scala [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
+
 import scala.meta.internal.implementation.TextDocumentWithPath
 import scala.meta.io.AbsolutePath
 
 import org.eclipse.{lsp4j => l}
 
 trait CodeLens {
   def isEnabled: Boolean
+
+  @nowarn
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/codelenses/CodeLens.scala:13]: Can we specify what we are silencing with these?

File: metals/src/main/scala/scala/meta/internal/metals/formatting/IndentOnPaste.scala [hunk 6]
```diff
@@ -146,6 +147,7 @@ case class IndentOnPaste(userConfig: () => UserConfiguration)
       case (line, _) =>
         PastedLine.plainOrEmpty(normalizeSpacesAndTabs(line, opts))
     }
+    @nowarn("msg=The outer reference")
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/formatting/IndentOnPaste.scala:150]: So aren't we silencing this same warning in other places in the `build.sbt`, but specifically here as well? I find it a bit confusing if we silence it in some packages in the `build.sbt` and then in this file here. Could we had this place in there as well?

File: metals/src/main/scala/scala/meta/internal/parsing/Trees.scala [hunk 7]
```diff
@@ -103,6 +101,9 @@ final class Trees(
           case Parsed.Success(tree) =>
             trees(path) = tree
             List()
+          case _ =>
+            scribe.info("Received unexpected parser result")
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/parsing/Trees.scala:105]: Will this happen often? Should this be info or warn?

File: metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala [hunk 8]
```diff
@@ -469,6 +467,7 @@ object WorksheetProvider {
   }
   final case class MdocRef(scalaVersion: String, value: Mdoc)
 
+  // TODO
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala:470]: Leftover TODO?

File: tests/cross/src/main/scala/tests/BasePCSuite.scala [hunk 9]
```diff
@@ -107,9 +108,12 @@ abstract class BasePCSuite extends BaseSuite {
       snippetAutoIndent = false
     )
 
-  protected def extraDependencies(scalaVersion: String): Seq[Dependency] = Nil
+  @nowarn
```
Comment (ckipp01) [inline tests/cross/src/main/scala/tests/BasePCSuite.scala:111]: Can we specify what we are silencing with these two in here?

**Review by ckipp01 — COMMENTED** [review]
Nice to see all the cleanup! Good job on this! Couple questions/suggestions.

File: build.sbt (hunk 3 above)
Comment (kpodsiad) [reply build.sbt:283]: Yes, they're false positives. It's because:
```scala
  for {
    value <- opt
    if value
    result <- computeSth()
  } yield result
```
is being desugared more or less to the:
```scala
  opt
    .withFilter(value => value)
    .flatMap(value =>
      computeSth()
        .map(result => result)
    )
```
and it is true that `value` in flatmap is unused.

I could rewrite them but it's an inadequate amount of work compared to the value it brings.

Comment (ckipp01) [reply build.sbt:283]: Gotcha, thanks for the explanation.

File: metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala (hunk 8 above)
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala:470]: Oh, I forgot about it.
Yeah, it's for @tgodzik because he introduced these methods :D

File: metals/src/main/scala/scala/meta/internal/metals/formatting/IndentOnPaste.scala (hunk 6 above)
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/metals/formatting/IndentOnPaste.scala:150]: Good catch!
I started with `@nowarns` in files but it yielded a compilation error in mtags since `nowarn` was introduced in some specific version of scala and this one slipped through. 
I'll move it to the `build.sbt` as others.

File: build.sbt (hunk 4 above)
Comment (kpodsiad) [reply build.sbt:284]: Sorry, could you rephrase the second sentence?

File: metals/src/main/scala/scala/meta/internal/parsing/Trees.scala (hunk 7 above)
Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/parsing/Trees.scala]: I changed severity to the `error` and I logged the class's simple name. Maybe it's too cautious?

File: build.sbt (hunk 4 above)
Comment (ckipp01) [reply build.sbt:284]: Ha, yea sorry. I mean this will ignore any message that has `in method` in it right? If so, are there potentially multiple warnings that might have `in method` in it that we are ignoring when we specifically only want to ignore one? Just trying to make sure this isn't too broad.

Comment (kpodsiad) [reply build.sbt:284]: IIRC every message that comes from `ScaladocParser.scala` file and has `in method reportError` so it's a bit moe narrow. But I think I can change it to the `parameter (name1|name2) in method reportError` and this should be very precise.

File: metals/src/main/scala/scala/meta/internal/parsing/Trees.scala [hunk 10]
```diff
@@ -102,9 +100,14 @@ final class Trees(
             )
           case Parsed.Success(tree) =>
             trees(path) = tree
-            List()
+            List.empty
+          case other =>
+            scribe.error(
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/parsing/Trees.scala:105]: I don;t think we need to log it since it's not possible currently to get any other result.

File: metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala (hunk 8 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala:470]: Wait, but what would I need to do?

Comment (kpodsiad) [reply metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala:470]: what is the difference between both `worksheetScala3Adjustments` if they both ignore the second parameter? Why one is adding +1 to the position?

Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/worksheets/WorksheetProvider.scala:470]: The second parameter looks like it can be removed. The second method has the additiona Position mapping. It's +1 because we wrap worksheet input in an object, since scripts do not yet work for Scala 3

**Comment by kpodsiad** [comment]
@ckipp01 @dos65 suggestions applied, CI is green, the number of warnings is equal to zero.

**Review by ckipp01 — APPROVED** [review]
🚢 it.

**Comment by kpodsiad** [comment]
Fatal warnings for `2.13.8` on the CI, will it be useful or troublesome?
